### PR TITLE
:warning: Follow XDG Directory standard for config/data/... files

### PR DESCRIPTION
**Pull request by cwrau**
**What this PR does / why we need it**: This PR switches the hardcoded paths (`$HOME/.cluster-api`) to be [XDG](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) conform (`$XDG_CONFIG_HOME/cluster-api`).

This removes clutter in the users' home directory.

Open questions:
  - I tried adding backwards compatibility to the old path inside the `reader_viper.go` but I'm not sure if this works for everything
  - The script `cmd/clusterctl/hack/create-local-repository.py` creates a `dev-repository` at `$HOME/.cluster-api/dev-repository`. I see three solutions to this;
    - The script uses an [xdg library](https://github.com/srstevenson/xdg), which has to be installed on the users' system, which we'll mention in the docs
    - The script does everything by itself, without needing dependencies
      - This could be a lot of work, especially for Windows and macOS, as I can't test these systems
    - The documentation will mention to overwrite `XDG_CONFIG_HOME` while using the script, this could be weird

    I personally vote for the first possibility

### CODE REVIEW

**Comment by cwrau** [comment]
The apidiff is failing, but that is expected, what can I do?

**Comment by killianmuldoon** [comment]
The api diff isn't merge blocking if the maintainers think this is worth merging without an API revision - the check is really informing rather than blocking for API changes so don't worry about trying to pass it :)

**Comment by killianmuldoon** [comment]
The api diff isn't merge blocking if the maintainers think this is worth merging without an API revision - the check is really informing rather than blocking for API changes so don't worry about trying to pass it - it would be good if you could sign the CLA though 🙂

**Comment by cwrau** [comment]
> The api diff isn't merge blocking if the maintainers think this is worth merging without an API revision - the check is really informing rather than blocking for API changes so don't worry about trying to pass it - it would be good if you could sign the CLA though slightly_smiling_face

Yes, my chef is working on the CLA 👍️

**Comment by cwrau** [comment]
CLA has been accepted 👍️

**Comment by killianmuldoon** [comment]
Good to hear! I'm not sure what you're looking for with the help tag though :laughing:

**Comment by cwrau** [comment]
> Good to hear! I'm not sure what you're looking for with the help tag though 😆

I just wanted to add the `help-wanted` label, but apparently that doesn't work this way 🤣

**Comment by killianmuldoon** [comment]
I think we normally only add that label to issues where we want someone to take them up - are you looking to have someone else pick up this PR? If so it might be better to create an issue describing the goals, link this draft PR and then add the help wanted tag.

**Comment by cwrau** [comment]
Ah, ok, then I misunderstood the point of that label 😅

No, I can keep this PR 👍

**Comment by sbueringer** [comment]
@cwrau Is the PR ready for review? (just asking because we usually use Draft/WIP to ~ signal implementation is still in progress and no reviews are required yet)

cc @ykakarap

**Comment by cwrau** [comment]
@sbueringer

Ah, I see, yeah, in that case it's still a draft, as I need to have the questions answered 👍

By anyone, really 😅

**Review by ykakarap — COMMENTED** [review]
>Ah, I see, yeah, I still need to have the questions answered 👍
>
>By anyone, really 😅

I will take a look at this.

**Review by ykakarap — COMMENTED** [review]
Looks like according to XDG documentation if `XDG_CONFIG_HOME` is not defined it should default to `$HOME/.config`. Meaning, we will effectively move from `$HOME/.cluster-api` to `$HOME/.config/cluster-api` in the general case. 

I do not have strong opinions on doing this move.  

Regarding your first question - I think establishing backward compatibility with `$HOME/.cluster-api` should be possible. This would also mean unless we break the contract both the `XDG_CONFIG_HOME` and `$HOME/.cluster-api` will need to be supported.

Regarding your second question - Are you suggesting that we move the `dev-repository` location also to `XDG_CONFIG_HOME`? If so, It might be easier to just build the support into the script we have. Looks like we dont need all of xdg library to begin with and can just have the logic that reads the `XDG_CONFIG_HOME` env var and default to the appropriate value if it is not defined. This would be simpler and would reduce external dependencies. As you pointed out I am not sure if this will be complicated for other OS environments. We will need more information before we can decide.

> The documentation will mention to overwrite XDG_CONFIG_HOME while using the script, this could be weird

I am not sure why a user will have to override `XDG_CONFIG_HOME`. Can you please explain?

**Comment by dims** [comment]
prior art from `kubectl` https://github.com/kubernetes/kubernetes/pulls?q=is%3Apr+XDG

**Comment by sbueringer** [comment]
cc @fabriziopandini

**Comment by vincepri** [comment]
@cwrau Still interested in working on this?

**Comment by richardcase** [comment]
I'm favour this change to using the XDG standard. It would be easy to have backwards compatibility for users with the current `~/.cluster-api`

@cwrau would you like to continue working on this?

**Comment by killianmuldoon** [comment]
@cwrau - could you change this from a draft to an open PR? I don't think this is getting reviews because draft can mean WIP - and definitely does when I see this in my notifications.

File: cmd/clusterctl/client/config.go [hunk 1]
```diff
@@ -134,7 +134,7 @@ type GetClusterTemplateOptions struct {
 	ClusterName string
 
 	// KubernetesVersion to use for the workload cluster. If unspecified, the value from os env variables
-	// or the .cluster-api/clusterctl.yaml config file will be used.
+	// or the $XDG_CONFIG_HOME/clusterctl.yaml or .cluster-api/clusterctl.yaml config file will be used.
```
Comment (elmiko) [inline cmd/clusterctl/client/config.go:137]: should this be `XDG_CONFIG_HOME/cluster-api/clusterctl.yaml` ?

**Review by elmiko — COMMENTED** [review]
this generally looks ok to me, but i have a concern about breaking our osx and windows users, also a nit inline, and do we need to update https://github.com/kubernetes-sigs/cluster-api/blob/main/cmd/clusterctl/hack/create-local-repository.py as well?

File: cmd/clusterctl/client/config.go (hunk 1 above)
Comment (cwrau) [reply cmd/clusterctl/client/config.go:137]: Yeah, you're right 👍️

**Comment by cwrau** [comment]
The tests that are failing are checks that verify that the path is `$HOME/.cluster-api` 😅

File: cmd/clusterctl/client/config/reader_viper.go [hunk 2]
```diff
@@ -58,14 +58,18 @@ func injectConfigPaths(configPaths []string) viperReaderOption {
...
 // newViperReader returns a viperReader.
-func newViperReader(opts ...viperReaderOption) Reader {
+func newViperReader(opts ...viperReaderOption) (Reader, error) {
+	configDirectory, err := xdg.ConfigFile(ConfigFolder)
+	if err != nil {
+		return nil, err
+	}
 	vr := &viperReader{
-		configPaths: []string{filepath.Join(homedir.HomeDir(), ConfigFolder)},
+		configPaths: []string{configDirectory, filepath.Join(xdg.Home, ConfigFolder)},
```
Comment (richardcase) [inline cmd/clusterctl/client/config/reader_viper.go:67]: I tested this with a config file in the olf location and its not getting picked up. In the logs i see:

```
No default config file available
```

I think what you need it the following change:

```go
const (
	// ConfigFolder defines the name of the config folder under $XDG_CONFIG_HOME.
	ConfigFolder = "cluster-api"
	// ConfigFolderLegacy defines the name of the old config folder under $HOME.
	ConfigFolderLegacy = ".cluster-api"
....
	vr := &viperReader{
		configPaths: []string{configDirectory, filepath.Join(xdg.Home, ConfigFolderLegacy)},
	}

```

With that change it picks up the file in the old location:

```
Using configuration File="/home/richard/.cluster-api/clusterctl.yaml"
```

**Comment by richardcase** [comment]
@cwrau  - with the open questions in the PR description, the first one is answered. How about the second question?

**Comment by cwrau** [comment]
> this generally looks ok to me, but i have a concern about breaking our osx and windows users, also a nit inline, and do we need to update [main/cmd/clusterctl/hack/create-local-repository.py](https://github.com/kubernetes-sigs/cluster-api/blob/main/cmd/clusterctl/hack/create-local-repository.py?rgh-link-date=2023-03-09T17%3A41%3A09Z) as well?

> @cwrau - with the open questions in the PR description, the first one is answered. How about the second question?

I guess we would add a dependency to the `create-local-repository.py` script with a `requirements.txt`, if no one minds that?

**Comment by richardcase** [comment]
@cwrau - with the `pull-cluster-api-apidiff-main` failure, it reminded me that we shouldn't change the value of this exported constant. To ensure that the go api hasn't changed, we'd need to put the value back and use a different constant for the new folder name. Sorry, i should've realized this before suggesting the name of the constant.

So, it would need to be changed to something like:

```go
// ConfigFolder defines the name of the config folder under $HOME.
ConfigFolder = ".cluster-api"
// ConfigFolderXDG defines the name of the config folder under $XDG_CONFIG_HOME.
ConfigFolderXDG = "cluster-api"
```

Also would you be able to squash your commits?

**Comment by cwrau** [comment]
The last `lint` errors are the following;

```
cmd/clusterctl/client/repository/overrides_test.go:25: File is not `gci`-ed with --skip-generated -s standard,default (gci)

cmd/clusterctl/client/repository/overrides.go:27: File is not `gci`-ed with --skip-generated -s standard,default (gci)

cmd/clusterctl/cmd/root.go:27: File is not `gci`-ed with --skip-generated -s standard,default (gci)
	"github.com/adrg/xdg"

cmd/clusterctl/cmd/root.go:29: File is not `gci`-ed with --skip-generated -s standard,default (gci)
	"github.com/MakeNowJust/heredoc"

cmd/clusterctl/cmd/root.go:31: File is not `gci`-ed with --skip-generated -s standard,default (gci)
	"github.com/spf13/cobra"

cmd/clusterctl/client/config/reader_viper.go:31: File is not `gci`-ed with --skip-generated -s standard,default (gci)
```

and

```
controlplane/kubeadm/main.go:153:2: SA1019: rand.Seed has been deprecated since Go 1.20 and an alternative has been available since Go 1.0: Programs that call Seed and then expect a specific sequence of results from the global random source (using functions such as Int) can be broken when a dependency changes how much it consumes from the global random source. To avoid such breakages, programs that need a specific result sequence should use NewRand(NewSource(seed)) to obtain a random generator that other packages cannot access. (staticcheck)
	rand.Seed(time.Now().UnixNano())
	^
bootstrap/kubeadm/main.go:145:2: SA1019: rand.Seed has been deprecated since Go 1.20 and an alternative has been available since Go 1.0: Programs that call Seed and then expect a specific sequence of results from the global random source (using functions such as Int) can be broken when a dependency changes how much it consumes from the global random source. To avoid such breakages, programs that need a specific result sequence should use NewRand(NewSource(seed)) to obtain a random generator that other packages cannot access. (staticcheck)
	rand.Seed(time.Now().UnixNano())
	^
main.go:208:2: SA1019: rand.Seed has been deprecated since Go 1.20 and an alternative has been available since Go 1.0: Programs that call Seed and then expect a specific sequence of results from the global random source (using functions such as Int) can be broken when a dependency changes how much it consumes from the global random source. To avoid such breakages, programs that need a specific result sequence should use NewRand(NewSource(seed)) to obtain a random generator that other packages cannot access. (staticcheck)
	rand.Seed(time.Now().UnixNano())
```

which I don't know how to fix 😅

File: cmd/clusterctl/client/config/reader_viper_test.go [hunk 3]
```diff
@@ -192,7 +192,7 @@ func Test_viperReader_GetWithoutDefaultConfig(t *testing.T) {
 
 	_ = os.Setenv("FOO_FOO", "bar")
 
-	v := newViperReader(injectConfigPaths([]string{dir}))
+	v, _ := newViperReader(injectConfigPaths([]string{dir}))
```
Comment (richardcase) [inline cmd/clusterctl/client/config/reader_viper_test.go:195]: nit: Guessing we expect an error not to have occured as well?

**Comment by sbueringer** [comment]
@cwrau I think running `make lint-fix` should fix them (not 100% sure)

**Comment by cwrau** [comment]
> @cwrau I think running make lint-fix should fix them (but not 100% sure)

Fixed the `gci` errors 👍️, but not the `rand.seed` errors 😅

**Comment by sbueringer** [comment]
> > @cwrau I think running make lint-fix should fix them (but not 100% sure)
> 
> Fixed the `gci` errors 👍️, but not the `rand.seed` errors 😅

Cluster API is still on Go 1.19, which is why we don't have 1.20 linter issues fixed on main. I think you can ignore those findings (I triggered the linter action, let's see if it's green)

**Comment by richardcase** [comment]
With the fallback to the old location, from my side:

/lgtm

**Comment by richardcase** [comment]
Also, from my side still:

/lgtm

File: cmd/clusterctl/hack/create-local-repository.py [hunk 4]
```diff
@@ -102,26 +103,28 @@ def load_providers():
...
 
-def get_home():
-    return os.path.expanduser('~')
 
 def get_repository_folder():
-    home = get_home()
-    return os.path.join(home, '.cluster-api', 'dev-repository')
+    config_dir = os.getenv("XDG_CONFIG_HOME")
+    if config_dir is None:
+        raise Exception('XDG_CONFIG_HOME variable is not set')
```
Comment (killianmuldoon) [inline cmd/clusterctl/hack/create-local-repository.py:125]: IMO this should probably be defaulted - according to the spec: https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

` $XDG_CONFIG_HOME defines the base directory relative to which user-specific configuration files should be stored. If $XDG_CONFIG_HOME is either not set or empty, a default equal to $HOME/.config should be used. `

File: cmd/clusterctl/cmd/root.go [hunk 5]
```diff
@@ -78,7 +68,11 @@ var RootCmd = &cobra.Command{
 			// version check is disabled. Return early.
 			return nil
 		}
-		output, err := newVersionChecker(configClient.Variables()).Check()
+		checker2, err := newVersionChecker(configClient.Variables())
```
Comment (killianmuldoon) [inline cmd/clusterctl/cmd/root.go:71]:
```suggestion
		checker, err := newVersionChecker(configClient.Variables())
```

File: cmd/clusterctl/cmd/root.go [hunk 6]
```diff
@@ -56,16 +56,6 @@ var RootCmd = &cobra.Command{
...
-		// Check if Config folder (~/.cluster-api) exist and if not create it
-		configFolderPath := filepath.Join(homedir.HomeDir(), config.ConfigFolder)
-		if _, err := os.Stat(configFolderPath); os.IsNotExist(err) {
-			if err := os.MkdirAll(filepath.Dir(configFolderPath), os.ModePerm); err != nil {
-				return errors.Wrapf(err, "failed to create the clusterctl config directory: %s", configFolderPath)
-			}
-		}
-		return nil
-	},
 	PersistentPostRunE: func(cmd *cobra.Command, args []string) error {
```
Comment (killianmuldoon) [inline cmd/clusterctl/cmd/root.go:59]: Is this behaviour retained in `xdg.ConfigFile`? It seems so but just wnat to make sure.

**Review by killianmuldoon — COMMENTED** [review]
@cwrau thanks for sticking with this! I think we're on the home stretch now.
The changes look good to me - just a couple of nits and a question about defaulting.

**Comment by killianmuldoon** [comment]
/retitle :warning: Follow XDG Directory standard for config/data/... files

Just want to ensure this is signalled in release notes etc.

File: cmd/clusterctl/cmd/root.go (hunk 6 above)
Comment (cwrau) [reply cmd/clusterctl/cmd/root.go:59]: Yes, the library will create the folder if necessary 👍️

**Comment by elmiko** [comment]
did those machinepool files get pulled in on the rebase?

**Comment by fabriziopandini** [comment]
lgtm pending CI green

**Comment by killianmuldoon** [comment]
@cwrau - regarding the broken links check - I think that should be fixed if you rebase on main.

**Comment by richardcase** [comment]
Looks like with a rebase this is good to go.

**Comment by sbueringer** [comment]
Thx! lgtm (from a high-level review)

**Review by killianmuldoon — COMMENTED** [review]
/lgtm
/approve

Thanks again for following through on this!

**Comment by cwrau** [comment]
🥳🎉

**Pull request merged**